### PR TITLE
Adds a skill system

### PR DESCRIPTION
**Pull request by nmajask**
# Document the changes in your pull request

**New PR because I needed to make it its own branch, so look at #13255 for older feedback. You can also make PRs for this branch if you are a nerd.**

Adds a skill system which is heavily inspired by Baystation, various RGPs and some of the implementation by Hopek in #10802.
There are 23 skills, each of which will be mentioned in a list below, and each skill possesses 5 levels, with each consecutive level being more difficult or costly to obtain. 
Players will have a pool of points with which to purchase these skills, and differing jobs will have differing base proficiencies (ex. Scientist will have a better starting Design skill than an Assistant). These skills and levels will have varying effects on relevant actions, and/or can offer bonuses depending on the situation.

## New Skills
Obviously, this adds new skills, 23 to be exact. Most are useless right now, but I do plan on fixing that later.

- Strength: How good you are at lifting, pushing, or carrying heavy objects.
- Dexterity: How good you are at quickly and accurately moving yourself and your limbs.
- Endurance: How good you are at tanking through hits and resisting chemicals and alcohol.
- Botany: How good you are at planting and tending to plants.
- Cooking: How good you are using culinary tools.
- Creativity: How creative you can be.
- Survival: How well you can survive by yourself.
- Piloting: How good you are at piloting mechs or other vehicles.
- Leadership: How good you are at leading and inspiring others.
- Hand-to-Hand: How good you are at fighting with your bare hands.
- Melee Weapons: How good you are at fighting with a toolbox or batlong.
- Ranged Weapons: How good you are at fighting with a gun.
- Forensics: How good you are at finding evidence from the environment.
- Medicine: How good you are at identifying and treating injuries.
- Anatomy: How good you are with the innards of humans.
- Chemistry: How good you are at mixing and identifying chemicals.
- Psychology: How good you are at identifying the feelings of others.
- Design: How good you are at researching new designs.
- Information Technology: How good you are at using and creating computers.
- Biology: How knowledgeable you are at understanding how life works.
- Mechanics: How good you are at creating and manipulating machines.
- Electronics: How good you are at creating and manipulating electronics.
- Atmospherics: How good you are at identifying and manipulating gasses and atmospheric tools.

## Why?

- Makes power gaming dependent on code instead of the interpretation of a rule by admins
- Allows for more interesting characters
- Increases RP

## To-Do

- [ ] Make the skills useful
- [ ] Balance stuff
- [ ] Fix some of the UI
- [ ] Get better sprites for the items and skill button
- [ ] Clean up some of this code spaghetti 

## Current Uses For Skills

- [ ] Strength: None.
- [ ] Dexterity: None.
- [ ] Endurance: None.
- [ ] Botany: None.
- [ ] Cooking: Some crafting recipes.
- [ ] Creativity: None.
- [ ] Survival: Using some tools and in some crafting recipes.
- [ ] Piloting: None.
- [ ] Leadership: None.
- [ ] Hand-to-Hand: None.
- [ ] Melee Weapons: None.
- [ ] Ranged Weapons: Accuracy for most guns.
- [ ] Forensics: Using forensic scanner.
- [x] Medicine: Using medical scanner, identifying wounds, and some surgeries.
- [ ] Anatomy: Most surgeries.
- [ ] Chemistry: Increases the speed of transferring reagents.
- [ ] Psychology: None.
- [ ] Design: Some crafting recipes.
- [ ] Information Technology: None.
- [ ] Biology: None.
- [ ] Mechanics: Using some tools and in some crafting recipes.
- [ ] Electronics: Using some tools and in some crafting recipes.
- [ ] Atmospherics: None.

## Screenshots

image
image

# Wiki Documentation

Will probably need a whole new page for this, plus some information about skill proficiency's in the appropriate job page. 

# Changelog

:cl:  
rscadd: Added a new skill system
tweak: Scanning someone with a health analyzer requires a Basic medical skill and takes time if your skill level is bellow Experienced.
tweak: Transferring reagents now takes time depending on your chemistry level.
tweak: Wounds now require certain skills to be identified or give detains when inspected.
tweak: Surgeries now have an increased chance to fail if your level in the required skill is low.
tweak: Tools are now slower if your level in the required skill is low and faster if its high.
/:cl:

### CODE REVIEW

File: code/datums/wounds/burns.dm [hunk 1]
```diff
@@ -126,32 +126,40 @@
...
+				if(WOUND_INFECTION_MODERATE to WOUND_INFECTION_SEVERE)
+					condition += ", <span class='deadsay'>with early signs of infection.</span>"
+				if(WOUND_INFECTION_SEVERE to WOUND_INFECTION_CRITICAL)
+					condition += ", <span class='deadsay'>with growing clouds of infection.</span>"
+				if(WOUND_INFECTION_CRITICAL to WOUND_INFECTION_SEPTIC)
+					condition += ", <span class='deadsay'>with streaks of rotten, pulsating infection!</span>"
+				if(WOUND_INFECTION_SEPTIC to INFINITY)
+					return span_deadsay("<B>[victim.p_their(TRUE)] [limb.name] is a mess of charred skin and infected rot!</B>")
+				else
+					condition += "!"
```
Comment (tattax) [inline code/datums/wounds/burns.dm]:
```suggestion
		if(skill_check(user, identifing_skill, descriptive_skilllevel))
			switch(infestation)
				if(WOUND_INFECTION_MODERATE to WOUND_INFECTION_SEVERE)
					condition += ", [span_deadsay("with early signs of infection.")]"
				if(WOUND_INFECTION_SEVERE to WOUND_INFECTION_CRITICAL)
					condition += ", [span_deadsay("with growing clouds of infection.")]"
				if(WOUND_INFECTION_CRITICAL to WOUND_INFECTION_SEPTIC)
					condition += ", [span_deadsay("with streaks of rotten, pulsating infection!")]"
				if(WOUND_INFECTION_SEPTIC to INFINITY)
					return span_deadsay("[span_bold([victim.p_their(TRUE)] [limb.name] is a mess of charred skin and infected rot!)]")
				else
					condition += "!"
```

File: code/modules/clothing/gloves/color.dm [hunk 2]
```diff
@@ -178,33 +178,24 @@
...
+	desc = "Cheap sterile gloves made from latex. Gives a tiny amout of protection."
 	icon_state = "latex"
 	item_state = "lgloves"
 	siemens_coefficient = 0.3
 	permeability_coefficient = 0.01
 	item_color="mime"
 	transfer_prints = TRUE
 	resistance_flags = NONE
-	var/carrytrait = TRAIT_QUICK_CARRY
+	armor = list("melee" = 0, "bullet" = 0, "laser" = 0, "energy" = 0, "bomb" = 0, "bio" = 5, "rad" = 0, "fire" = 0, "acid" = 5)
```
Comment (tattax) [inline code/modules/clothing/gloves/color.dm]:
```suggestion
	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 0, BIO = 5, RAD = 0, FIRE = 0, ACID = 5)
```

File: code/modules/clothing/gloves/color.dm [hunk 3]
```diff
@@ -178,33 +178,24 @@
...
-
-/obj/item/clothing/gloves/color/latex/equipped(mob/user, slot)
-	..()
-	if(slot == SLOT_GLOVES)
-		ADD_TRAIT(user, carrytrait, CLOTHING_TRAIT)
-
-/obj/item/clothing/gloves/color/latex/dropped(mob/user)
-	..()
-	REMOVE_TRAIT(user, carrytrait, CLOTHING_TRAIT)
+	armor = list("melee" = 0, "bullet" = 0, "laser" = 0, "energy" = 0, "bomb" = 0, "bio" = 10, "rad" = 0, "fire" = 0, "acid" = 10)
```
Comment (tattax) [inline code/modules/clothing/gloves/color.dm]:
```suggestion
	armor = list(MELEE = 0, BULLET = 0, LASER = 0, ENERGY = 0, BOMB = 0, BIO = 10, RAD = 0, FIRE = 0, ACID = 10)
```

File: code/modules/mob/living/carbon/human/human.dm [hunk 4]
```diff
@@ -956,35 +956,28 @@
...
-						return
-				else
-					buckle_mob(target, TRUE, TRUE, 90, 1, 0)
-		visible_message(span_warning("[src] fails to fireman carry [target]!"))
-	else
+	if(!can_be_firemanned(target) || incapacitated(FALSE, TRUE))
 		to_chat(src, span_warning("You can't fireman carry [target] while they're standing!"))
+		return
+	visible_message(span_notice("[src] starts lifting [target] onto their back.."), span_notice("You start to lift [target] onto your back..."))
+	if(!do_after(src, 50, TRUE, target, required_skill = SKILL_STRENGTH, required_skill_level = SKILLLEVEL_BASIC, skill_delay_mult_scaling = list(SKILLLEVEL_UNSKILLED = 4, SKILLLEVEL_BASIC = 2, SKILLLEVEL_TRAINED = 1, SKILLLEVEL_EXPERIENCED = 0.5, SKILLLEVEL_MASTER = 0.25)))
```
Comment (tattax) [inline code/modules/mob/living/carbon/human/human.dm]:
```suggestion
	if(!do_after(src, 5 SECONDS, target, TRUE, required_skill = SKILL_STRENGTH, required_skill_level = SKILLLEVEL_BASIC, skill_delay_mult_scaling = list(SKILLLEVEL_UNSKILLED = 4, SKILLLEVEL_BASIC = 2, SKILLLEVEL_TRAINED = 1, SKILLLEVEL_EXPERIENCED = 0.5, SKILLLEVEL_MASTER = 0.25)))
```

File: code/modules/projectiles/guns/ballistic.dm [hunk 5]
```diff
@@ -251,7 +258,7 @@
 				to_chat(user, span_notice("\The [src]'s [bolt_wording] is already cocked!"))
 			return
 		bolt_locked = FALSE
-	if (user)
+	if (user && do_after(user, 0, target = src, stayStill = FALSE, required_skill = required_skill, required_skill_level = required_skill_level, skill_delay_mult_scaling = skill_reload_delay))
```
Comment (tattax) [inline code/modules/projectiles/guns/ballistic.dm]:
```suggestion
	if (user && do_after(user, 0, src, stayStill = FALSE, required_skill = required_skill, required_skill_level = required_skill_level, skill_delay_mult_scaling = skill_reload_delay))
```

File: code/modules/projectiles/guns/ballistic.dm [hunk 6]
```diff
@@ -278,7 +285,7 @@
 	if(!istype(AM, mag_type))
 		to_chat(user, span_warning("\The [AM] doesn't seem to fit into \the [src]..."))
 		return FALSE
-	if(user.transferItemToLoc(AM, src))
+	if(user.transferItemToLoc(AM, src) && do_after(user, 0, target = src, stayStill = FALSE, required_skill = required_skill, required_skill_level = required_skill_level, skill_delay_mult_scaling = skill_reload_delay))
```
Comment (tattax) [inline code/modules/projectiles/guns/ballistic.dm]:
```suggestion
	if(user.transferItemToLoc(AM, src) && do_after(user, 0, src, stayStill = FALSE, required_skill = required_skill, required_skill_level = required_skill_level, skill_delay_mult_scaling = skill_reload_delay))
```

File: code/modules/projectiles/guns/ballistic.dm [hunk 7]
```diff
@@ -344,21 +351,22 @@
...
-				to_chat(user, span_notice("You load [num_loaded] [cartridge_wording]\s into \the [src]."))
-				playsound(src, load_sound, load_sound_volume, load_sound_vary)
-				if(can_reload_say)
-					user.say(reload_say, forced = "reloading")
-				if (chambered == null && bolt_type == BOLT_TYPE_NO_BOLT)
-					chamber_round()
-				A.update_icon()
-				update_icon()
-			return
+			if(do_after(user, 0, target = src, stayStill = FALSE, required_skill = required_skill, required_skill_level = required_skill_level, skill_delay_mult_scaling = skill_reload_delay))
```
Comment (tattax) [inline code/modules/projectiles/guns/ballistic.dm]:
```suggestion
			if(do_after(user, 0, src, stayStill = FALSE, required_skill = required_skill, required_skill_level = required_skill_level, skill_delay_mult_scaling = skill_reload_delay))
```

File: code/modules/reagents/reagent_containers/glass.dm [hunk 8]
```diff
@@ -494,4 +496,4 @@
 	icon_state = "urn_closed"
 	amount_per_transfer_from_this = 0 // No reagent transfer allowed, it's closed
 	possible_transfer_amounts = list(0)
-	to_chat(user, "<span class = 'notice'>You close the lid of [src] and lock it.</span>")
\ No newline at end of file
+	to_chat(user, "<span class = 'notice'>You close the lid of [src] and lock it.</span>")
```
Comment (tattax) [inline code/modules/reagents/reagent_containers/glass.dm]:
```suggestion
	to_chat(user, span_notice("You close the lid of [src] and lock it."))
```

File: yogstation/code/modules/skills/appraiser.dm [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+	desc = "A device that looks up the qualifications of the scanned being and assesses the best job for them."
+	icon = 'icons/obj/device.dmi'
+	icon_state = "appraiser"
+	var/use_roundstart = TRUE
+	var/show_skill_groups = TRUE
+
+/obj/item/appraiser/attack(mob/living/M, mob/living/carbon/human/user)
+	flick("[icon_state]-scan", src)
+  
+	if(!do_after(user, 10, target = M, required_skill = SKILL_LEADERSHIP, skill_delay_mult_scaling = list(SKILLLEVEL_UNSKILLED = 2, SKILLLEVEL_BASIC = 1.5, SKILLLEVEL_TRAINED = 1, SKILLLEVEL_EXPERIENCED = 0.5, SKILLLEVEL_MASTER = 0)))
```
Comment (tattax) [inline yogstation/code/modules/skills/appraiser.dm:12]:
```suggestion
	if(!do_after(user, 1 SECONDS, M, required_skill = SKILL_LEADERSHIP, skill_delay_mult_scaling = list(SKILLLEVEL_UNSKILLED = 2, SKILLLEVEL_BASIC = 1.5, SKILLLEVEL_TRAINED = 1, SKILLLEVEL_EXPERIENCED = 0.5, SKILLLEVEL_MASTER = 0)))
```

File: yogstation/code/modules/skills/skill_trainer.dm [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+	if(destroy_after && (!change_skills || !change_skills.len))
+		qdel(src)
+	return TRUE
+
+/obj/item/skilltrainer/attack(mob/living/M, mob/user)
+	if(!can_train(M, user))
+		return
+	playsound(src, 'sound/weapons/slice.ogg', 5)
+	if(!do_after(user, 10, target = M) || !can_train(M, user))
```
Comment (tattax) [inline yogstation/code/modules/skills/skill_trainer.dm:41]:
```suggestion
	if(!do_after(user, 1 SECONDS, M) || !can_train(M, user))
```

File: code/__HELPERS/mobs.dm [hunk 11]
```diff
@@ -282,10 +282,24 @@ GLOBAL_LIST_EMPTY(species_list)
...
 
-/proc/do_after(mob/user, delay, atom/target = null, needhand = TRUE, progress = TRUE, datum/callback/extra_checks = null, stayStill = TRUE)
+/proc/do_after(mob/user, delay, atom/target = null, needhand = TRUE, progress = TRUE, datum/callback/extra_checks = null, stayStill = TRUE, required_skill = null, required_skill_level = SKILLLEVEL_UNSKILLED, list/skill_delay_mult_scaling = list())
 	if(!user)
 		return FALSE
 	var/atom/target_loc = null
+	if(!isnull(required_skill) && usesSkills(user))
+		var/datum/skill/user_skill = find_skill(user, required_skill)
+		if(skill_check(user, required_skill, required_skill_level))
+			if(skill_delay_mult_scaling.len > 0)
```
Comment (tattax) [inline code/__HELPERS/mobs.dm]:
```suggestion
			if(skill_delay_mult_scaling.len)
```

File: code/datums/wounds/burns.dm (hunk 1 above)
Comment (nmajask) [reply code/datums/wounds/burns.dm]: You can just make a PR with the changes in the future

File: code/__HELPERS/mobs.dm (hunk 11 above)
Comment (tattax) [reply code/__HELPERS/mobs.dm]: if you want to i can commit these directly @nmajask

Comment (nmajask) [reply code/__HELPERS/mobs.dm:292]: > if you want to i can commit these directly @nmajask

Make a PR with the changes so I can just merge that @tattax

**Comment by maxion12345** [comment]
y'know maybe we could condense some of these skills
medical and anatomy could probably be one skill, so doctors won't have to spec into two skills to be useful (leaving other skills with less) (doctors are nerds tho)

Also not sure if it was considered yet but what about species? Do all species get the same points?
Humans could get the most, lizards get less but they get strength bonuses, moth get less but get dexterity bonus, etc

**Comment by MusselMain** [comment]
> y'know maybe we could condense some of these skills medical and anatomy could probably be one skill, so doctors won't have to spec into two skills to be useful (leaving other skills with less) (doctors are nerds tho)
> 
> Also not sure if it was considered yet but what about species? Do all species get the same points? Humans could get the most, lizards get less but they get strength bonuses, moth get less but get dexterity bonus, etc

The issue with that, is that Anatomy is a skill detectives should have to preform autopsies, and if we combine the two it might be a significantly stronger then other skills (unless we combine other skills to each other as well)

**Comment by ilikesaying** [comment]
image
WHEEEEEEEEEEEZE

**Review by Altoids1 — APPROVED** [review]
I think this is cool and good. I think it'd be less work and less balance arguments to reduce the number of skills by a fair bit, but that's not really a dealbreaker for me.

**Pull request closed without merging**